aqt/generate_combinations.py

```diff
@@ -0,0 +1,335 @@
+        + "\n]"
+        for key, item_formatter in (
+            ("qt", fmt_dict_entry),
+            ("tools", fmt_dict_entry),
+            ("modules", fmt_module_entry),
+        )
+    ] + [
+        f'"{key}": [\n  ' + fmt_version_list(combos[key], depth=1) + "\n]"
+        for key in ("versions", "new_archive")
+    ]
+
+    return "[{" + ", ".join(root_element_strings) + "}]"
+
+
+def compare_combos(
+    actual_combos, expected_combos, actual_name, expect_name, printer: Callable
+):
+    list_of_str_keys = "versions", "new_archive"
+
+    def compare_modules_entry(actual_mod_item, expect_mod_item):
+        version = actual_mod_item["qt_version"]
+        actual_modules, expect_modules = set(actual_mod_item["modules"]), set(
+            expect_mod_item["modules"]
+        )
+        mods_missing_from_actual = expect_modules - actual_modules
+        mods_missing_from_expect = actual_modules - expect_modules
+        if mods_missing_from_actual:
+            printer(
+                f"{actual_name}['modules'] for Qt {version} is missing {mods_missing_from_actual}"
+            )
+        if mods_missing_from_expect:
+            printer(
+                f"{expect_name}['modules'] for Qt {version} is missing {mods_missing_from_expect}"
+            )
+
+    def to_set(a_list: Union[List[str], List[Dict]]) -> Set:
+        if len(a_list) == 0:
+            return set()
+        if isinstance(a_list[0], str):
+            return set(a_list)
+        assert isinstance(a_list[0], Dict)
+        return set([str(a_dict) for a_dict in a_list])
+
+    def print_difference(superset, subset, subset_name, key):
+        missing_from_superset = sorted(superset - subset)
+        if not missing_from_superset:
+            return
+        printer(f"{subset_name}['{key}'] is missing these entries:")
+        if key in list_of_str_keys:
+            printer(format(missing_from_superset))
+            return
+        for el in missing_from_superset:
+            printer(format(el))
+
+    for root_key in actual_combos.keys():
+        print(f"\nComparing {root_key}:\n{'-' * 40}")
+        if root_key == "modules":
+            for actual_row, expect_row in zip(
+                actual_combos[root_key], expected_combos[root_key]
+            ):
+                assert actual_row["qt_version"] == expect_row["qt_version"]
+                compare_modules_entry(actual_row, expect_row)
+            continue
+
+        actual_set = to_set(actual_combos[root_key])
+        expected_set = to_set(expected_combos[root_key])
+        print_difference(expected_set, actual_set, actual_name, root_key)
+        print_difference(actual_set, expected_set, expect_name, root_key)
+
+
+def alphabetize_modules(combos: Dict):
+    for i, item in enumerate(combos["modules"]):
+        combos["modules"][i]["modules"] = sorted(item["modules"])
+
+
+if __name__ == "__main__":
+    Settings.load_settings()
+    logger = logging.getLogger("aqt.generate_combos")
```

Comment:
When introduce a new logger, please add it in logging.ini definition.
Current logger structure is

```
root -
       - aqt
                - main
                - archives
                - helper
                - installer
                - list
                - updater
```

Reply:
Oops, good call!

@miurahr, I have a question about the loggers used in `aqt/metadata.py`. The `metadata.show_list` function uses the `aqt.list` logger, but `metadata.MetadataFactory` still uses the `aqt.archives` logger. All of the other loggers appear to have the same name as the file in which they are used. Would it be appropriate to rename the `aqt.list` logger to `aqt.metadata`, and use that logger in both `metadata.MetadataFactory` and `metadata.show_list`?

Reply:
> I have a question about the loggers used in aqt/metadata.py. The metadata.show_list function uses the aqt.list logger, but metadata.MetadataFactory still uses the aqt.archives logger.

This inconsistency is just a bug, sorry.